src/osd/PG.h

```diff
@@ -2395,10 +2403,75 @@ class PG : DoutPrefixProvider {
   virtual void agent_delay() = 0;
   virtual void agent_clear() = 0;
   virtual void agent_choose_mode_restart() = 0;
+
+  virtual bool can_handle_fast(OpRequestRef& op) = 0;
 };
 
 ostream& operator<<(ostream& out, const PG& pg);
 
 ostream& operator<<(ostream& out, const PG::BackfillInterval& bi);
 
+// for no pg lock completion
+enum COMP_ACK_TYPE {
+  OP_COMP_ACK,
+  OP_COMP_ACK_COMP,
+};
+enum COMP_OP_TYPE {
+  OP_COMP_PRIMARY_OP,
+  OP_COMP_REP_OP,
+  OP_COMP_NO_OP,
+};
+
+class CompletionItem {
+public:
+  /* completion lock */
+  Mutex comp_lock;
+
+// need PG lock
+  /* update_peer_last_complete_ondisk() from REPOP_REPLY */
+  pg_shard_t fromosd;
+  eversion_t lcod;
+  
+  /* last_update_applied */
+  eversion_t applied_version;
+
+  /* callback func which need pg lock */
+  list<Context*> on_applied_with_pg_lock;
+
+  /* Replica versions in MSG_OSD_REPOP */
+  eversion_t rep_committed_version;
+  eversion_t rep_applied_version;
+  int ackerosd;
+
+  /* From OnOpCommit, OnOpApply */
+  pg_shard_t shard;
+  version_t epoch;
+
+  /* RepGather */
+  void * rep_gather;
+
+// don't need PG lock
+  ceph_tid_t rep_tid;
+  int comp_type; // 1: primary 2: replica
+  OpRequestRef op;
+
+  bool all_committed;
+  bool all_applied;
+  bool completed;
+  bool do_reserve_check_completion;
+  utime_t now;
+
+  CompletionItem(int comp_type): comp_lock("CompItem::comp_lock"), rep_gather(NULL), comp_type(comp_type) {
+    all_committed = false;
+    all_applied = false;
+    completed = false;
+    do_reserve_check_completion = false;
+    ackerosd = 0;
+  }
+
+  void register_on_applied_with_pg_lock(Context *c) {
+    if (!c) return;
+    on_applied_with_pg_lock.push_back(c);
+  }
```

Comment:
Let me show a concrete example about my suggestion. please see below sequence (note that this is MSG_OSD_REPOP).
Current code dose not use repop_queue and RepGather in the case of handling MSG_OSD_REPOP. However i would like to reuse them in order to make simple code. 
repop_queue is used for checking completed OPs as well as MSG_OSD_OP case.

A. when a write request is issued (Rep Op case)
 1. Receive MSG_OSD_REPOP  2. PG lock (acquire) -> 3. add repop_queue -> 4. sub_op_modify() -> 4. PG lock (release)

B. when commit, apply are completed (Rep Op case)
 1. complete commit or apply. -> 2. Repop(OP) lock (acquire) -> 3. If all commit or apply are finished, send ack to Primary OSD -> 5. Repop(OP) lock (release) -> 5. delay.. (6 will happen 1 second later in the batching manner)-> 6. queue_for_completion (see D.)

D. queue_for_completion (update PG metadata such as lcod) (Rep Op case)
 1. PG lock (acquire) -> 2. do_lazy_completion (iterate repop_queue for lazy completion and remove repop from repop_queue) -> 3. PG lock (release)